index.js

```diff
@@ -1,3 +1,216 @@
+   * large it will break it up and put the remainding chunks
+   * into a queue.
+   * NOTE: You can only write new chunks once the callback is
+   * called with no errors.
+   *
+   * @param {AudioBuffer} chunk (or Buffer) containing the data to be output
+   * @param {Function} callback callback with error and chunk parameters
+   * @return void
+   * @api public
+   */
+  function write (chunk, callback) {
+    if (!callback) callback = noop
+
+    if (options._closed) return callback(new Error('Write cannot occur after the speaker is closed.'))
+
+    if (chunk && options._busy) {
+      callback(new Error('Cannot write chunk as the buffer was busy.'))
+    }
+
+    next(chunk, null, callback)
+
+    function next (chunk, remainder, callback) {
+      if (options.handler) {
+        options._busy = true
+
+        var chunkBuf = isAudioBuffer(chunk) ? Buffer.from(pcm.toArrayBuffer(chunk, options)) : chunk || new Buffer(0)
+        var remainderBuf = isAudioBuffer(remainder) ? Buffer.from(pcm.toArrayBuffer(remainder, options)) : remainder || new Buffer(0)
+
+        var queue = Buffer.concat([remainderBuf, chunkBuf], remainderBuf.length + chunkBuf.length)
+
+        var output = queue.length > options.chunkSize ? queue.slice(0, options.chunkSize) : queue
+        var remaining = queue.length > options.chunkSize ? queue.slice(options.chunkSize, queue.length) : new Buffer(0)
+
+        binding.write(options.handler, output, output.length, (written) => {
+          if(!remaining.length < 1) {
+            next(null, remaining, callback)
+          } else {
+            if (options.autoFlush && remaining.length < options.chunkSize) {
+              binding.flush(options.handler, function (success) {
+                if (success != 1) {
+                  options._busy = false
+                  callback(new Error('Could not flush the audio output.'))
+                } else {
+                  options._busy = false
+                  callback(null)
+                }
+              })
+            } else {
+              options._busy = false
+              callback(null)
+            }
+          }
+        })
+      } else {
+        callback(new Error('Could not write remaining chunks as the speaker is closed.'))
+      }
+    }
+  }
+
+  function sink (callback) {
+    return audioSink((data, callback) => {
+      if (options._closed) return callback(true)
+      setTimeout(callback, options.samplesPerFrame / options.sampleRate)
+    })
+  }
+
+  /**
+   * The end function closes the speaker and stops
+   * it from writing anymore data. The output data
+   * that was already written can be optionally flushed.
+   * NOTE: You cannot write anymore data after closing the speaker.
+   *
+   * @param {Boolean} flush flushes the written data (default is false)
+   * @param {Function} callback callback with error parameter
+   * @return void
+   * @api public
+   */
+  function end (flush, callback) {
```

Comment:
I [mistook a bug](https://github.com/audiojs/web-audio-write/issues/) I saw previously in the review for something completely different, which I will address here instead now.  Sorry for the confusion! @dfcreative 

So, I don't think we need `flush` or `callback` here, especially because it adds complications between the Node browser implementation. 

 1. [Conventions on functional packages](https://github.com/audiojs/contributing/wiki/Streams-convention#sinkdestinationwritable)  does not have parameters for `write.end()`, so we should probably stay true to this.
 2. Flushing should be a default in `audio-speaker`, as it is the expected result here in my opinion.  If people want more fine control over writing to the output device, maybe we can point them directly to [`audio-mpg123`](https://github.com/audiojs/audio-mpg123)?
 3. `binding.flush` isn't actually asynchronous.  It uses a callback in order to send information to you (this is a confusing setup inherited from `speaker` package, that [would be replaced with the NAPI branch](https://github.com/audiojs/audio-mpg123/pull/10) in the future)

I hope this made more sense, and provides a good case.  :smile: